test/db_connection/metrics_test.exs

```diff
@@ -0,0 +1,126 @@
+defmodule MetricsTest do
+  use ExUnit.Case, async: false
+
+  defmodule DummyDB do
+    use DBConnection
+
+    def start_link(opts), do: DBConnection.start_link(__MODULE__, opts)
+
+    def connect(_opts), do: {:ok, %{connected: true}}
+    def disconnect(_reason, _state), do: :ok
+    def checkout(state), do: {:ok, state}
+    def checkin(_, state), do: {:ok, state}
+    def handle_execute(_request, _state, _opts), do: {:ok, %{result: "dummy result"}}
+
+    def handle_begin(_, _), do: :ok
+    def handle_close(_, _, _), do: :ok
+    def handle_commit(_, _), do: :ok
+    def handle_deallocate(_, _, _, _), do: :ok
+    def handle_declare(_, _, _, _), do: :ok
+    def handle_fetch(_, _, _, _), do: :ok
+    def handle_prepare(_, _, _), do: :ok
+    def handle_rollback(_, _), do: :ok
+    def handle_status(_, _), do: :ok
+    def handle_execute(_, _, _, _), do: :ok
+    def ping(_), do: :ok
+  end
+
+  defmodule DelayedDummyDB do
+    # Delegate to DummyDB for all functions, except for checkout/1
+    defdelegate start_link(opts), to: DummyDB
+    defdelegate connect(opts), to: DummyDB
+    defdelegate disconnect(reason, state), to: DummyDB
+    defdelegate checkin(state1, state2), to: DummyDB
+    defdelegate handle_execute(request, state, opts), to: DummyDB
+    defdelegate handle_begin(arg1, arg2), to: DummyDB
+    defdelegate handle_close(arg1, arg2, arg3), to: DummyDB
+    defdelegate handle_commit(arg1, arg2), to: DummyDB
+    defdelegate handle_deallocate(arg1, arg2, arg3, arg4), to: DummyDB
+    defdelegate handle_declare(arg1, arg2, arg3, arg4), to: DummyDB
+    defdelegate handle_fetch(arg1, arg2, arg3, arg4), to: DummyDB
+    defdelegate handle_prepare(arg1, arg2, arg3), to: DummyDB
+    defdelegate handle_rollback(arg1, arg2), to: DummyDB
+    defdelegate handle_status(arg1, arg2), to: DummyDB
+    defdelegate handle_execute(arg1, arg2, arg3, arg4), to: DummyDB
+    defdelegate ping(arg1), to: DummyDB
+
+    def checkout(state) do
+      :timer.sleep(1000)  # Simulate a delay
```

Comment:
Those tests are going to be very expensive, unfortunately. It may be this single test file is slower than the whole test suite. :( Would it be possible for us to use :connection_listeners to detect when all connections are up instead of relying on long :timer.sleep periods?

Reply:
Hm, I'm not sure, I'll take a look

Reply:
Alternatively, regular erlang tracing can be used.